harness/determined/searcher/search_runner.py

```diff
@@ -0,0 +1,328 @@
+                raise RuntimeError("validationCompleted event is invalid without a metric")
+
+            operations = self.search_method.on_validation_completed(
+                request_id,
+                event.validationCompleted.metric,
+            )
+
+        elif event.experimentInactive:
+            logging.info(
+                f"experiment {self.search_method.searcher_state.experiment_id} is "
+                f"inactive; state={event.experimentInactive.experimentState}"
+            )
+            if (
+                event.experimentInactive.experimentState
+                == determinedexperimentv1State.STATE_COMPLETED
+            ):
+                self.search_method.searcher_state.experiment_completed = True
+
+            raise _ExperimentInactiveException(event.experimentInactive.experimentState)
+
+        elif event.trialProgress:
+            logging.debug(
+                f"trialProgress({event.trialProgress.requestId}, "
+                f"{event.trialProgress.partialUnits})"
+            )
+            request_id = uuid.UUID(event.trialProgress.requestId)
+            self.search_method.searcher_state.trial_progress[request_id] = float(
+                event.trialProgress.partialUnits
+            )
+            progress = self.search_method.progress()
+            operations = [Progress(progress)]
+        else:
+            raise RuntimeError(f"Unsupported event {event}")
+        return operations
+
+    def run_experiment(
+        self,
+        experiment_id: int,
+        session: client.Session,
+        prior_operations: Optional[List[Operation]],
+    ) -> None:
+        experiment_is_active = True
+
+        try:
+            while experiment_is_active:
+                time.sleep(5)
+                events = self.get_events(session, experiment_id)
+                if events is None:
+                    continue
+                logging.info(json.dumps([SearchRunner._searcher_event_as_dict(e) for e in events]))
+                # the first event is an event we have already processed and told master about it
+                # however, we may not have saved the state after that event if we crashed
+                # after POSTing operations but before saving state
+                last_event_id = self.search_method.searcher_state.last_event_id
+                first_event = True
+                for event in events:
+                    if (
+                        first_event
+                        and last_event_id != 0
+                        and last_event_id >= event.id >= 0
+                        and prior_operations is not None
+                    ):
+                        logging.info(f"Resubmitting operations for event.id={event.id}")
+                        operations = prior_operations
+                    else:
+                        try:
+                            operations = self._get_operations(event)
+                        except _ExperimentInactiveException as ex:
```

Comment:
Just detect `event.experimentInactive` outside of `self._get_operations()`.



Using exceptions like this is a bit of a hack I think.  It's like changing your `_get_operations` signature to secretly be:

```python
def _get_operations(event) -> Union[List[Operations], _ExperimentInactiveException]:
```

because it actually has additional, expected behaviors other than returning a list of operations.  But when you use exceptions for those extra behaviors, it isn't part of the typing of the function.